src/gurt/telemetry.c

```diff
@@ -0,0 +1,1857 @@
+ * process.
+ */
+int
+d_tm_init(int rank, uint64_t memSize)
+{
+	char tmp[D_TM_MAX_NAME_LEN];
+	uint64_t *base_addr = NULL;
+	int rc = 0;
+	int buffLen = 0;
+
+	if (shmemRoot && root) {
+		D_INFO("d_tm_init already completed for rank %d\n", rank);
+		return rc;
+	}
+
+	shmemRoot = (uint64_t *)d_tm_allocate_shared_memory(rank, memSize);
+	if (shmemRoot != NULL) {
+		shmemIdx = (uint8_t *)shmemRoot;
+		shmemFree = memSize;
+		D_DEBUG(DB_TRACE, "Shared memory allocation success!\n"
+		       "Memory size is %"PRIu64" bytes at address 0x%"PRIx64
+		       "\n", memSize, (uint64_t)shmemRoot);
+		/*
+		 * Store the base address of the shared memory as seen by the
+		 * server in this first uint64_t sized slot.
+		 * Used by the client to adjust pointers in the shared memory
+		 * to its own address space.
+		 */
+		base_addr = (uint64_t *)d_tm_shmalloc(sizeof(uint64_t));
+		/*
+		 * Just allocated the pool, and this memory is dedicated
+		 * for this process, so a d_tm_shmalloc failure cannot
+		 * happen here.  Just checking for a null pointer for
+		 * good hygiene.
+		 */
+		if (!base_addr) {
+			rc = -DER_NO_SHMEM;
+			D_ERROR("Out of shared memory: %d", rc);
+			D_GOTO(failure, rc);
+		}
+		*base_addr = (uint64_t)shmemRoot;
+	} else {
+		rc = -DER_NO_SHMEM;
+		D_ERROR("Could not allocate shared memory: %d", rc);
+		D_GOTO(failure, rc);
+	}
+
+	root = (struct d_tm_node_t *)d_tm_shmalloc(sizeof(struct d_tm_node_t));
+
+	/*
+	 * Just allocated the pool, and this memory is dedicated
+	 * for this process, so a d_tm_shmalloc failure cannot
+	 * happen here.  Just checking for a null pointer for
+	 * good hygiene.
+	 */
+	if (!root) {
+		rc = -DER_NO_SHMEM;
+		D_ERROR("Out of shared memory: %d", rc);
+		D_GOTO(failure, rc);
+	}
+
+	sprintf(tmp, "rank %d", rank);
+	buffLen = strnlen(tmp, D_TM_MAX_NAME_LEN);
+	if (buffLen == D_TM_MAX_NAME_LEN) {
+		rc = -DER_EXCEEDS_PATH_LEN;
+		D_ERROR("Path name exceeds length: %d", rc);
+		D_GOTO(failure, rc);
+	}
+	buffLen += 1; /* make room for the trailing null */
+
+	root->name = (char *)d_tm_shmalloc(buffLen * sizeof(char));
+	if (!root->name) {
+		rc = -DER_NO_SHMEM;
+		D_ERROR("Out of shared memory: %d", rc);
+		D_GOTO(failure, rc);
+	}
+	strncpy(root->name, tmp, buffLen);
+
+	root->child = NULL;
+	root->sibling = NULL;
+	root->metric = NULL;
+	root->d_tm_type = D_TM_DIRECTORY;
+
+	/*
+	 * This mutex protects the d_tm_add_metric operations
+	 */
+	rc = D_MUTEX_INIT(&addlock, NULL);
+	if (rc != 0) {
+		D_ERROR("Mutex init failure: %d.", rc);
+		D_GOTO(failure, rc);
+	}
+
+	D_INFO("Telemetry and Metrics initialized for rank: %u\n", rank);
+	return rc;
+
+failure:
+	D_ERROR("Failed to initialize telemetry and metrics API: %d", rc);
+	return rc;
+}
+
+/*
+ * Releases resources claimed by init
+ * Currently, only detaches from the shared memory to keep the memory and the
+ * shared mutex objects available for the client to read data even when the
+ * producer has gone offline.
```

Comment:
I'm not sure how common this use case would be. I certainly wouldn't have the expectation that I would still be able to read telemetry from a process that has gone away.

When an ioserver restarts, does it reclaim the same segment it had before? If so, is that segment scrubbed so that there's no possibility of getting stale metrics from the previous process?

I guess I'm just a little wary of this idea of the segment having a lifetime independent of the server. How would a consumer know when the ioserver has restarted, for example?

Reply:
Thinking about it a bit more, I don't think we can rely on fini cleaning things up, because of course there will be crashes. Given that, it seems like focusing on the init/reclaim side of things would be best.

One thing that occurs to me as a sanity-check would be to write the current producer's PID into the segment and consumers could check to see if that process is running. Probably not ideal though because it's somewhat racy, but maybe better than nothing?

Reply:
My angle was that the client should be able to read the metrics even if the producer process goes away because that might provide useful data.  However, I don't think that ability is a must have feature.  I thought people would appreciate the decoupling though.

When an ioserver restarts, it does reclaim the same segment it had before, but with a different virtual address assigned to the new process.  When the segment is reinitialized, the base address is re-written.  When the client performs a read operation, the pointers are validated to make sure they point to an address within the client's virtual address space of the shared memory pool.  The address is calculated based upon the base address that was written by the ioserver.  Once that base address changes, any data that was written by the previous ioserver instance is invalidated implicitly, because all of the addresses used are now 'invalid'.  In the current scheme, they are valid addresses for the client up until the moment the new ioserver process loads, reattaches to that shared memory segment, and then finally writes its new base address.  There isn't an explicit wipe of the shared memory at that time because it wasn't necessary, but I can do that if there is concern about that.  

I worked at the design to hide the fact that the ioserver restarted.  That said, the client is not signaled that the ioserver process changed.  It could however, read the base address from the shared memory to compare what it had before and could determine that the ioserver process has changed.  Alternatively, I could add a signal of some sort.  It really depends on what we want to do with that information on the client side.  It's a transparent change at the moment.